File: Data Structures & Algorithms/longest-increasing-subsequence/submission-5.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        return helper(nums, 0, -1);
    }
    public int helper(int[] nums, int i, int prev){
        if (i == nums.length){
            return 0;
        }
        int res = helper(nums, i+1, prev); // No take
        if (prev == -1 || nums[prev] < nums[i]){
            res = Math.max(res, 1 + helper(nums, i+1, i));
        }
        return res;
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-5.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = [-1] * (n+1)
        def helper(i):
            if i == 0:
                return 1
            if i < 0:
                return 0
            if memo[i] != -1:
                return memo[i]
            memo[i] = helper(i-1) + helper(i-2)
            return memo[i]
        return helper(n)